### ISSUE TITLE
Integrate proxy detection into the verification flow

### ISSUE THREAD

**Issue by manuelwedler** [ethereum/sourcify#1712]
The PoC we implement in https://github.com/ethereum/sourcify/issues/1655 detects and resolves proxy contracts on the fly when our API is called. It would be nice from a data analysis perspective to mark contracts in our DB as proxies by already running the detection during verification. People using our parquet exports could make use of this data then. Furthermore, we could add EIP1167 contracts to our database (https://github.com/ethereum/sourcify/issues/1624).

The implementation resolution still needs to be done during an API call, because we cannot keep track of changed implementations. Only the implementations of EIP1167 contracts could be added to the database (but not necessary as it can be read out of the bytecode very quickly).

### Needed changes

- [ ] Specify DB schema change
- [ ] Implement DB schema change
- [ ] Integrate detection into the verification process (should be in `server` and not in `lib-sourcify`)
- [ ] Still run proxy resolution on the fly when the API is called
- [ ] Upgrade DB
- [ ] Migrate existing data by running proxy detection on all already verified contracts

**Comment by manuelwedler** [comment]
Two options for adding a new `proxy_type` column:

- add it to `contract_deployments` (or some other VerA table)
- add it to `sourcify_matches`

I think it's not clear where to put it inside a VerA table. It could also be `verified_contracts` or `compiled_contracts`. I mainly think `contract_deployments` is good because we currently run the detection on the onchain code, but that is also not set in stone. It should also work on the recompiled code.

Generally, I think the proxy detection is something more high level than the rest of the information we store in the database. It's rather a analytics or classification information and has nothing to do with the compilation or deployment info. Therefore, it does not feel correct to me to store it inside the current VerA tables. I tend to say that `sourcify_matches` is the best option at the moment for these reasons.

**Comment by kuzdogan** [comment]
My wanting to put this in the VerA tables is because it's a useful information for data processing etc. But I tend to agree that for now it seems better to place in `sourcify_matches`. Maybe we should add this in the Vera v2 draft issue for future discussion.

Also, does storing the proxy information add a benefit for the calls. Do we plan to cache the implementation address too or resolve on the fly each time? When we have the proxy type, do we say WhatsABI to not parse and investigate the bytecode but does it say "this is a OZ proxy so I'll directly do the storage slot RPC calls"?

**Comment by marcocastignoli** [comment]
> But I tend to agree that for now it seems better to place in sourcify_matches. Maybe we should add this in the Vera v2 draft issue for future discussion.

I agree with this

> Do we plan to cache the implementation address too or resolve on the fly each time?

The answer to this question is in the issue description:

> The implementation resolution still needs to be done during an API call, because we are not a block explorer and cannot keep track of changed implementations. Only the implementations of EIP1167 contracts could be added to the database (but not necessary as it can be read out of the bytecode very quickly).

**Comment by manuelwedler** [comment]
> Do we plan to cache the implementation address too or resolve on the fly each time?

My plan was to resolve it each time on the fly. How would you cache it? If just in the server, the benefit would be minimal, since we are running a lot of instances. We could do in the database if you think it makes sense. But then we would need to store more information like last time updated. In that case, I think it would even make sense to have its own table for the proxy information.

Now I'm even thinking this is quite a good idea. If we just reference the `deployment_id` in a `proxy_info` table, we don't even need to modify our existing tables. And it's tied to a deployment what makes perfect sense. The table could look like this:

```sql
CREATE TABLE proxy_info (
    deployment_id uuid, -- can be primary foreign key
    proxy_type varchar,
    implementation_address bytea,
    implementation_updated_at timestamptz
);
```

In that form, I would even be up for adding it to VerA. It doesn't introduce a change to an existing table that feels misplaced.

What do you think about this table?


> When we have the proxy type, do we say WhatsABI to not parse and investigate the bytecode but does it say "this is a OZ proxy so I'll directly do the storage slot RPC calls"?

We could potentially do this. Unfortunately, latest version introduced a change from that we cannot infer the slot purely on the proxy type (https://github.com/shazow/whatsabi/compare/v0.22.0...v0.22.1). We would also need to store the storage slot if we don't want to run the detection every time we run the resolution. Imo, storing the slot is not worth it, since the detection runs extremly quick. When playing around with whatsabi, the detection took only 0.44 ms on average (see https://github.com/sourcifyeth/data-analysis-scripts/blob/main/proxy-detection-results/durations-stats.json).

**Comment by kuzdogan** [comment]
> My plan was to resolve it each time on the fly. How would you cache it? If just in the server, the benefit would be minimal, since we are running a lot of instances. We could do in the database if you think it makes sense. But then we would need to store more information like last time updated. In that case, I think it would even make sense to have its own table for the proxy information.

Yeah let's keep it simple and resolve each time, shouldn't be a huge burden.

> If we just reference the deployment_id

I liked the additional table idea. Just, as above, maybe no need to save implementation address? 

If saving the proxy types does not give any tangible benefits and resolution is that fast, maybe we don't need to store anything at all? 😅 The only reason I see then is if anyone wants to see e.g. how many "OZ Proxies are there?" but they can also search the sources to some extent.

**Comment by manuelwedler** [comment]
> If saving the proxy types does not give any tangible benefits and resolution is that fast, maybe we don't need to store anything at all? 😅 The only reason I see then is if anyone wants to see e.g. how many "OZ Proxies are there?" but they can also search the sources to some extent.

You are right. The main reason to have it in the database is the analytics aspect. Imo, for that it's still worth it. Maybe it is also worth to store the storage slot actually for that reason, and then we can at least skip the detection.

Besides this, storing the proxy info in the database during verification would enable to store minimal proxies (EIP1167)

> I liked the additional table idea. Just, as above, maybe no need to save implementation address?

I added it as a caching mechanism, to save on the RPC calls. But that could also be interesting for analysis.

**Comment by kuzdogan** [comment]
> Besides this, storing the proxy info in the database during verification would enable to store minimal proxies (EIP1167), which we cannot verify because they don't have source code.

How do we do that in that case? We are not going to index chains. Are we going to check the bytecode for it being a EIP1167 when someone requests a certain contract?

Keep in mind you need to run the proxy detection retrospectively for already verified contracts.

I'm actually getting more skeptical if this is really needed and should have priority. Did you already start implementing anything?

**Comment by manuelwedler** [comment]
> How do we do that in that case? We are not going to index chains. Are we going to check the bytecode for it being a EIP1167 when someone requests a certain contract?

See https://github.com/ethereum/sourcify/issues/1624
I think the idea was to have some way to "verify" the minimal proxies through our API. We would just check the onchain bytecode for a provided chain and address, and store it directly in our db if we detect a minimal proxy. However, I think we might need a new API endpoint for that, and I'm skeptical now that we want that. Maybe we should not include the minimal proxies in here.

> I'm actually getting more skeptical if this is really needed and should have priority. Did you already start implementing anything?

No, I didn't implement anything. I still think this issue would be generally a nice thing to have (excluding the minimal proxies), but we can deprioritize it for now.

**Comment by kuzdogan** [comment]
What does @marcocastignoli think?

**Comment by marcocastignoli** [comment]
Let's postpone the implementation. Maybe a call would be helpful to do more planning for this.
